lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+  public:
+    DiagnoseInfiniteRecursion() {}
+
+  private:
+    static bool hasRecursiveCallInPath(SILBasicBlock &Block,
+                                       SILFunction *Target) {
+      // Process all instructions in the block to find applies that reference
+      // the parent function.  Also looks through vtables for statically
+      // dispatched (witness) methods.
+      auto *TargetModule = Target->getModule().getSwiftModule();
+      for (auto &I : Block) {
+        auto *AI = dyn_cast<ApplyInst>(&I);
+        if (AI && AI->getCalleeFunction() && AI->getCalleeFunction() == Target)
+          return true;
+
+        if (FullApplySite FAI = FullApplySite::isa(&I)) {
+          // Don't touch dynamic dispatch.
+          if (isa<ObjCMethodInst>(FAI.getCallee()))
+            continue;
+
+          auto &M = FAI.getModule();
+          if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+            auto ClassType = CMI->getOperand()->getType();
+
+            // FIXME: If we're not inside the module context of the method,
+            // we may have to deserialize vtables.  If the serialized tables
+            // are damaged, the pass will crash.
+            //
+            // Though, this has the added bonus of not looking into vtables
+            // outside the current module.  Because we're not doing IPA, let
+            // alone cross-module IPA, this is all well and good.
+            auto *BGC = ClassType.getNominalOrBoundGenericNominal();
+            if (BGC && BGC->getModuleContext() != TargetModule) {
+              continue;
+            }
+
+            auto *F = getTargetClassMethod(M, ClassType, CMI);
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+
+          if (auto *WMI = dyn_cast<WitnessMethodInst>(FAI.getCallee())) {
+            SILFunction *F;
+            SILWitnessTable *WT;
+
+            std::tie(F, WT) = M.lookUpFunctionInWitnessTable(
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
+                                            SILFunction *TargetFn,
+                                            const ExitBlockSet &ExitBlocks) {
+      enum RecursiveState {
+        FoundRecursion          = 0,
+        FoundPathOutOfFunction  = 1,
+      };
+
+      // Set the entry block into the state map.
+      llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+      States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
+
+      // Make the processing stack and seed it with the entry block.
+      SmallVector<SILBasicBlock *, 16> Stack;
+      Stack.push_back(Fn.getEntryBlock());
+
+      while (!Stack.empty()) {
```

Comment:
I don't understand why this is a worklist. It can be done with a single RPO pass over the blocks.

Reply:
Successor blocks of recursive functions aren't visited with the worklist approach, but are visited - and need to be rejected - by RPO.